extractors: restructuring classes

awillecke opened the pull request:
Restructuring of classes to remove duplicate code due to the introduction of SqlExtractor

* renamed old BaseExtractor to OmnetExtractor
* New BaseExtractor implements basic reading of SQLite files and category conversion
* SqlExtractor inherits from new BaseExtractor
* BaseExtractor can now include the filename in the returned DataFrame

awillecke committed:
extractors: restructuring classes

Restructuting of classes to remove duplicate code due to introduction of
SqlExtractor

* renamed old BaseExtractor to OmnetExtractor
* New BaseExtractor implements basic reading of SQLite files and
  category conversion
* SqlExtractor inherits from new BaseExtractor

hagau committed:
extractors: change YAML tag for `BaseExtractor`

hagau committed:
extractors: call the constructor of `BaseExtractor` when initializing
`OmnetExtractor`

hagau committed:
extractors: fix parameter names in call to
`BaseExtractor.read_sql_from_file`

hagau committed:
extractors: rewrite column data conversion to convert only explicitly
given columns

Instead of using a heuristic to determine whether a column
could usefully be converted to a categorical data type, convert only
columns explicitly named in the `categorical_columns` parameter to a
(automatically derived) categorical data type. Columns given in the
`numerical_columns` parameter will be converted to `float`, or to the
type given explicitly in a dictionary.

hagau committed:
exporters: don't try converting column data types when exporting data

hagau committed:
plots: adapt column data conversion in `PlottingReaderFeather` to the
explicit conversion

hagau committed:
examples: update the AoI example to use explicit categorical and
numerical column data conversion

hagau commented:
I've rewritten the `convert_columns_to_category` (and renamed it to the more fitting `convert_columns_dtype`) to only convert explicitly given column to either categorical or numerical data. 
- The types of columns that are not mentioned are chosen by pandas automatically
- numerical columns can be converted to a specified type via a given dictionary
- `examples/AoI.yaml` has been updates as a more complex example
- unfortunately, upon reading the data back in in `PlottingReaderFeather`, the conversion has to be done again to save on memory usage. I've added the two parameters there and the example in `AoI.yaml` uses anchors and aliases to alleviate the need to repeat the same unwieldy data structure.
- I've only tested the `PatternMatchingBulkExtractor` via the AoI example, but I'm fairly confident everything else should work too. I'll test the other extractors while cleaning up the recipes :slightly_smiling_face:

hagau committed:
examples: change SqlExtractor example to use explicit column data
conversion

awillecke commented on exporters.py:
```diff
@@ -141,8 +141,7 @@ def prepare_concatenated(self, data_list, job_list):
             concat_result = dask.delayed(pd.concat)(map(operator.itemgetter(0), data_list), ignore_index=True)
-            convert_columns_result = dask.delayed(RawExtractor.convert_columns_to_category)(concat_result)
-            job = dask.delayed(self.save_to_disk)(convert_columns_result, self.output_filename, self.format)
+            job = dask.delayed(self.save_to_disk)(concatenate, self.output_filename, self.format)
```
I think this should be `concat_result` instead of `concatenate` right?

hagau replied:
Yes, thank you! Sorry, autocomplete tricked me there and I only tested the `prepare_separated` pathway :blush: 
I've pushed a fix now.

awillecke commented:
Thank you for implementing explicit type conversion! It already works great. Currently, I'm testing a few recipes and will report back.

hagau committed:
exporters: fix typo

hagau committed:
examples: update lineplot recipe with explicit categorical/numerical
column definitions

awillecke committed:
extractor: arbitrary regex for input_files

allow the regex for input_files used to find files by a given
regex in path or file. Pervious implementation only supported
regex for files in a given directory.

awillecke commented:
I've tested the `SqlExtractor` and `RawExtractor` and both work as expected :+1: 

Also, I've added the possibility to have regex in every part of files added as `input_files` to allow something like this when data is structured in subdirectories:
```
evaluation: !Evaluation
  extractors:
  - dataset: !SqlExtractor
      input_files: !!python/list
        - "/tmp/ns3/experiment_[a-z].*/scenario_[a-z].*/run_[0-9].*.db"
```

hagau committed:
data_io: generalize the regular expression that is used to match on a
non-regex path prefix

hagau committed:
data_io: return the common root path instead of the regex-free prefix

hagau committed:
data_io: convert docstrings for regex path matching functions to
numpydoc format

hagau commented:
@awillecke Thanks, that's way more useful than the rather limited functionality that existed before.
I've generalized the used regular expression to allow for paths like `/tmp/ns3/experiment._[a-z].*/scenario_[a-z].*/run_[0-9].*.db` (note the first dot). This will still match paths with e.g. escaped parentheses as part of the regex, but that will not result in any problems in the selection.
I've also converted the docstrings to numpydoc format. I will add a comment in the documentation that it is the format for docstrings used throughout the code.

hagau commented:
> @awillecke Thanks, that's way more useful than the rather limited functionality that existed before. I've generalized the used regular expression to allow for paths like `/tmp/ns3/experiment._[a-z].*/scenario_[a-z].*/run_[0-9].*.db` (note the first dot). This will still match paths with e.g. escaped parentheses as part of the regex, but that will not result in any problems in the selection. I've also converted the docstrings to numpydoc format. I will add a comment in the documentation that it is the format for docstrings used throughout the code.

I have to add that for things like escaped parentheses in the path, one has to use a plain or single-quoted scalar in YAML. Unfortunately, I've used the double-quoted style in most examples. I'm going to add a comment in the documentation and a couple of examples in the example recipes to make that more clear.

awillecke commented:
Thank you for review. I've not considered that case.

awillecke committed:
extractor: convert types after applying tags

hagau committed:
extractors: fix order of categorical column data

Previously the order was determined automatically, which often lead to a
rather useless or wrong order due to lexical sorting. Instead, use a key
extractor to split each individual entry into tuples of alphabetical and
numerical elements and sort those tuples for a more proper ordering.

awillecke commented:
IMO it is intuitive to set all columns in the `categorical_columns` and `numerical_columns` parameters that are expected to be the output of the extractor.  Also, sometimes a user would like to convert tags to other data types. 

Then it makes sense to call `convert_columns_dtype` after the whole data frame is built (so after applying the tags). Hence, I moved the call for type conversion to `read_query_from_file` which now extracts data via `read_sql_from_file` and applies tags in the case of `OmnetExtractor` first, before setting the respective types.

hagau committed:
exporters: add the column dtype conversion back to the
`FileResultProcessor`

Since a transform can add arbitrary columns, allow explicit conversion
when exporting a DataFrame.

hagau commented:
> IMO it is intuitive to set all columns in the `categorical_columns` and `numerical_columns` parameters that are expected to be the output of the extractor. Also, sometimes a user would like to convert tags to other data types.
> 
> Then it makes sense to call `convert_columns_dtype` after the whole data frame is built (so after applying the tags). Hence, I moved the call for type conversion to `read_query_from_file` which now extracts data via `read_sql_from_file` and applies tags in the case of `OmnetExtractor` first, before setting the respective types.

Thanks, that makes more sense.

hagau commented:
I've extended the column dtype conversion to apply a more sensible ordering on the elements. Previously, the ordering was automatically derived and lexical for strings (e.g. 'pdr_3' came after 'pdr_29'), now a relatively simple heuristic is used to determine the type contained in a string, it's split into tuples and those tuples are then used to produce a proper order.

I've also added back the type conversion when exporting to allow for conversion after a transform. One could argue that the transform are a better place for that, since the point of the conversion is mostly to conserve memory, but that would be a separate PR, I think. The call to the conversion function is also much cheaper now, so multiple calls to it for different columns are not a problem.

hagau commented:
I will test and adapt more of the example recipes later on, but I think this is ready to be merged :+1:

awillecke commented:
I also think it's fine to merge now :+1:
